tests/org.dataflowanalysis.analysis.tests/src/org/dataflowanalysis/analysis/tests/dfd/MicroSecEndTest.java

```diff
@@ -0,0 +1,248 @@
+
+                if (hasNodeCharacteristic(node, "Stereotype", "local_logging") && !hasNodeCharacteristic(node, "Stereotype", "log_sanitization")) {
+                    addToMap(violationsMap, variant, 11, node);
+                    violation = true;
+                }
+
+                if (hasNodeCharacteristic(node, "Stereotype", "logging_server") && !hasDataCharacteristic(node, "Stereotype", "message_broker")) {
+                    addToMap(violationsMap, variant, 12, node);
+                    violation = true;
+                }
+                return violation;
+            });
+        }
+    }
+    
+    @Disabled
+    @ParameterizedTest
+    @ValueSource(strings = {
+        "anilallewar",
+        "apssouza22"/*,
+        "callistaenterprise",
+        "ewolff",
+        "ewolff-kafka",
+        "fernandoabcampos",
+        "georgwittberger",
+        "jferrater",
+        "koushikkothagal",
+        "mdeket",
+        "mudigal-technologies",
+        "rohitghatol",
+        "spring-petclinic",
+        "sqshq",
+        "yidongnan"*/
+    })
+    void testConstraints(String model) {
+        Map<Integer, Map<Integer, List<AbstractVertex<?>>>> violationsMap = new HashMap<>();
+        for (int variant : TUHH_MODELS.get(model)) {
+            String variationName = model + "_" + variant;
+            System.out.println(variationName);
+            analysis(Paths.get(location, model, variationName)
+                    .toString(),variant,violationsMap);
+        }
+        for (int variant : violationsMap.keySet()) {
+            System.out.println("Variant: " + variant);
+            System.out.println("Violations: " + violationsMap.get(variant)
+                    .keySet());
+            assertFalse(violationsMap.get(variant)
+                    .keySet()
+                    .contains(variant));
+        }
+    }
+
+    private boolean hasNodeCharacteristic(AbstractVertex<?> node, String type, String value) {
+        return node.getAllVertexCharacteristics()
+                .stream()
+                .anyMatch(n -> n.getTypeName()
+                        .equals(type)
+                        && n.getValueName()
+                                .equals(value));
+    }
+
+    private boolean hasDataCharacteristic(AbstractVertex<?> node, String type, String value) {
+        return node.getAllDataCharacteristics()
+                .stream()
+                .anyMatch(v -> v.getAllCharacteristics()
+                        .stream()
+                        .anyMatch(c -> c.getTypeName()
+                                .equals(type)
+                                && c.getValueName()
+                                        .equals(value)));
+    }
+
+    private void addToMap(Map<Integer, Map<Integer, List<AbstractVertex<?>>>> map, int variant, int rule, AbstractVertex<?> node) {
+        map.putIfAbsent(variant, new HashMap<>());
+
+        Map<Integer, List<AbstractVertex<?>>> secondaryMap = map.get(variant);
```

Comment:
What does the map contain? The variable name should be somewhat indicative of what is inside